nservicebus/messaging/evolving-contracts.md

```diff
@@ -1,108 +1,177 @@
 
-### Breaking down large contract assemblies
 
-In the early days of a system, combining all events, commands and messages into a single contracts assembly might be a good place to start. However, as the system grows, breaking down the contracts into smaller parts makes it easier to evolve a system safely. An obvious reason is a rise in the number of subscribers for events published by a specific endpoint. It's not desirable to expose commands that are meant to be consumed by a single receiver to all the subscribers interested in that same endpoint's events.
+--------------------------------------------------------------------
+Different proposal:
+
+## Techniques
+
+There are different techniques to evolve message contracts, each bringing it's own advantages and disadvantages. When selecting a technique it's important to carefully plan the migration strategy, the needs of message receivers/subscribers and message senders/publishers.
+
+### New message type
+
+Ship multiple versions of a message contract in the same assembly by creating a new message type for the new contract version. The new type typically indicates its relationship via its name, e.g.
+
+```
+// first version of the message contract
+public class CreateOrder : ICommand {
+    ...
+}
+
+// new version of the message contract
+public class CreateOrderV2 : Icommand{
+    ...
+}
+```
+
+* Consumers can be updated to the new contract version independent from updating the contract assembly, giving more flexibility in planning the upgrade process.
+* Allows more flexibility about the contract changes as no type-level compatibility has to be ensured.
+
+When updating commands/messages:
+* Ensure the receiver of the message has a handler for both types to ensure it can process messages that are still in the queue or coming from endpoints still using the old contract.
+
+When updating events:
+* Update all subscribers to have handlers (and subscribe) to both events before changing the publisher.
+
+### Use inheritance
+
+Ship multiple versions of a message contract in the same assembly by using inheritance. The new contract version inherits from the previous version, gaining all its parent's properties.
+
+```
+// first version of the message contract
+public class OrderCreatedEvent : IEvent {
+    ...
+}
+
+public class OrderCreatedEventV2 : OrderCreatedEvent {
+    ...
+}
+```
+
+* Can only be used to make additive changes
+* Flattening the inheritance chain / removing older versions is more difficult compared to the other options
+* Something about multiple/all message handlers being invoked
+* Something about routing complexity?
+
+When updating commands/messages:
+* Update all senders to the new version before adding a new message handler for the new version. This might delay the time till the new message contract can be implemented by the receiver significantly.
```

Comment:
```suggestion
When updating commands/messages, update all senders to the new version before adding new message handlers for the new version to receivers. This might delay the time till the new message contract can be implemented by receivers significantly.
```